### ISSUE TITLE
callable union: none of those signatures are compatible with each other

### ISSUE THREAD

**Issue by KotlinIsland** [microsoft/TypeScript#45269]
# Bug Report

function union doesn't work the same way as function interface

### 🔎 Search Terms

function type union

### 🕗 Version & Regression Information
4.4.0-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&allowUnreachableCode=true&allowUnusedLabels=true&noUnusedLocals=true&noUnusedParameters=true&jsx=0&noFallthroughCasesInSwitch=true&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true&ts=4.4.0-beta#code/KYDwDg9gTgLgBAYwgOwM71AQwLZgDbABccA3gLABQcAPAIIB8AFAJTEtwC89cAbhAJYATOHAD0ouADNMeVMDjJgAc0wx+PInADktLXH6o4g4AjyYowYQCMArvH4xDPGTfkGFwDVDgXMggHSUNLQANHAAQkysvAKClAC+nHDsXMnMnNw2yMaS-IpxFJSgkLCIKOhwWLgEAEyEQXAAPsl0URlp7XxCzA3NjHRhkSncXYI9VEnDHalZOXmWlIsUVfjA1Mg22FbAUFEsRSA4q+ub21Bh6FB5SlFL4nAAKgAW7sUWqKj8KPqGyBDwCBkZisBECFHuAFFMAgnnBsMAtjs4BBJHAYE95Fkvsg0QBPMDyLT9BhTKajdJ9AYRNqpcl6J6YQyfJTIVQ2d5hWzwP6KZGo9EQORwZmsmDs4CGczyJC4VT8EHyADuDlhwGhsP+GKgYJWtROiN2LHYByOeo2BouMCuyBu+woQA)

### 💻 Code

```ts
export const example: {
 <A>(): () => void  // false negative: 'A' is declared but its value is never read.
 <A, B>(): void
} = () => () => undefined

export const example2:
  | (<A>() => () => void)
  | (<A, B>() => void)
  = () => () => undefined

example<number>()()
example<number, string>()

// This expression is not callable.
// Each member of the union type '(<A>() => () => void) | (<A, B>() => void)' has signatures, but none of those signatures are compatible with each other.
example2<number>()() 
example2<number, string>()
```

### 🙁 Actual behavior

has type error

### 🙂 Expected behavior
it works

**Comment by fatcerberus** [comment]
`example` is a type with two call signatures, i.e. overloads, which is internally represented with an *intersection*, not a union.  A union of function types means "I have a function that has one of these signatures, but I don't know which" (i.e. the same thing union means in any other context), therefore you can only call it with a set of parameters that satisfies *all* the members of the union.  TS is telling you, there is no possible call you can write that's compatible with both function types and so won't let you call it.
